### PR TITLE
New Function - Get-WslDistributionOnline

### PR DESCRIPTION
**Pull request by JediNite**
Created a new function to retrieve the WSL distributions that are available from online sources.

Also incremented the version to 2.0.1.

### CODE REVIEW

File: Wsl.psm1 [hunk 1]
```diff
@@ -272,6 +272,34 @@ function Get-WslDistribution
...
 
+<#
+.EXTERNALHELP
+Wsl-help.xml
+#>
+# Retrieves listing of distributions available from online sources
+function Get-WslDistributionOnline {
+    [CmdletBinding()]
+    param(
+    )
```
Comment (SvenGroot) [inline Wsl.psm1]: Please follow the same style as the existing functions.

```suggestion
    param()
```

File: Wsl.psm1 [hunk 2]
```diff
@@ -272,6 +272,34 @@ function Get-WslDistribution
...
+
+    process {
```
Comment (SvenGroot) [inline Wsl.psm1]: A process block is not necessary for a function that takes no pipeline input.

File: Wsl.psm1 [hunk 3]
```diff
@@ -272,6 +272,34 @@ function Get-WslDistribution
...
+        $store = $false
+        Invoke-Wsl "--list", "--online" -IgnoreErrors | Select-Object -Skip 1 | ForEach-Object {
```
Comment (SvenGroot) [inline Wsl.psm1]: The skip seems unnecessary since you're checking for the header by name.

File: Wsl.psm1 [hunk 4]
```diff
@@ -272,6 +272,34 @@ function Get-WslDistribution
...
+            $name, $friendlyName = $_ -split ' ', 2
+            if ($store -eq $true ) {
```
Comment (SvenGroot) [inline Wsl.psm1]: Please follow the style of the rest of the file.

```suggestion
            if ($store) {
```

File: Wsl.psm1 [hunk 5]
```diff
@@ -272,6 +272,34 @@ function Get-WslDistribution
...
+                $friendlyName = $($friendlyName.Split(@(" "), [System.StringSplitOptions]::RemoveEmptyEntries)) -join " "
+                [PSCustomObject]@{
+                    "Name"         = $name
+                    "FriendlyName" = $friendlyName
+                }
+            }
+            if ( $name -contains "NAME" ) {
```
Comment (SvenGroot) [inline Wsl.psm1]: Please follow the style of the rest of the file (no space inside the parentheses, and add a blank line after the closing brace).

Additionally, this seems safe because the "NAME" string doesn't appear to be localized, but we should have a comment about that. The text before it is localized, so better to use a case-sensitive exact match to make sure it's not in that text.

```suggestion

            # The "NAME", "FRIENDLY NAME" header is not localized so can be used to find the start
            # of the list
            if ($name -ceq "NAME") {
```

File: Wsl.psm1 [hunk 6]
```diff
@@ -272,6 +272,34 @@ function Get-WslDistribution
...
+    [CmdletBinding()]
+    param(
+    )
+
+    process {
+        $store = $false
+        Invoke-Wsl "--list", "--online" -IgnoreErrors | Select-Object -Skip 1 | ForEach-Object {
+            $name, $friendlyName = $_ -split ' ', 2
+            if ($store -eq $true ) {
+                $friendlyName = $($friendlyName.Split(@(" "), [System.StringSplitOptions]::RemoveEmptyEntries)) -join " "
```
Comment (SvenGroot) [inline Wsl.psm1]:
```suggestion
                $friendlyName = $friendlyName.Trim()
```

File: Wsl.psm1 [hunk 7]
```diff
@@ -272,6 +272,34 @@ function Get-WslDistribution
...
+                [PSCustomObject]@{
```
Comment (SvenGroot) [inline Wsl.psm1]: Please create a custom class at the top of the file and return that instead, to be consistent with other functions that return data in this module. Make sure it matches the name of the Output type shown in the external help file you added.

File: Wsl.psd1 [hunk 8]
```diff
@@ -9,7 +9,7 @@
 RootModule = 'Wsl.psm1'
 
 # Version number of this module.
-ModuleVersion = '2.0.0'
+ModuleVersion = '2.0.1'
```
Comment (SvenGroot) [inline Wsl.psd1]: This module uses semver, so this should be 2.1.0 since it adds new functionality.
```suggestion
ModuleVersion = '2.1.0'
```

File: Wsl.psm1 (hunk 5 above)
Comment (SvenGroot) [reply Wsl.psm1]: Actually, make this an `elseif` since there's no need to do this check if $store is already true.

File: Wsl.Tests.ps1 [hunk 9]
```diff
@@ -230,6 +230,12 @@ Describe "WslManagementPS" {
         Test-Distro ($distros | Where-Object { $_.Name -eq "wslps_raw" }) "wslps_raw" 2 "Stopped"
     }
 
+    It "Can list online distributions" {
+        # Tests that there are online distributions available
+        $distros = Get-WslDistributionOnline
+        $distros.Count | Should -BeGreaterThan 0
```
Comment (SvenGroot) [inline Wsl.Tests.ps1]: Thanks for adding a test! This can be simplified a bit:


```suggestion
        $distros | Should -Not -BeNullOrEmpty
```

I think it would also be a good idea to make sure the list does not contain any items with an empty Name or FriendlyName. This should guard against a premature match on the NAME string in the localized text at the top of the command.

File: README.md [hunk 10]
```diff
@@ -316,6 +316,36 @@ This cmdlet wraps the functionality of `wsl.exe --shutdown`.
 There is no benefit to using this over `wsl.exe --shutdown`. It is provided purely for the sake of
 completionism.
 
+### Get-WslDistributionOnline
+
+The [`Get-WslDistributionOnline`][] cmdlet gets information about WSL distributions available from online sources.
```
Comment (SvenGroot) [inline README.md]: Please use a maximum line width of 100 characters in Markdown files (with some leeway for long links).

File: ChangeLog.md [hunk 11]
```diff
@@ -1,5 +1,9 @@
 # What's new in the WSL Management PowerShell module
 
+## Version 2.0.1 (2023-09-26)
+
+- Added support for retrieving WSL distribution listing from online sources via [`Get-WslDistributionOnline`][] cmdlet.
```
Comment (SvenGroot) [inline ChangeLog.md]: Please use a maximum line width of 100 characters in Markdown files.

File: docs/Get-WslDistributionOnline.md [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+
+## SYNTAX
+
+```
+Get-WslDistributionOnline [<CommonParameters>]
+```
+
+## DESCRIPTION
+
+The `Get-WslDistributionOnline` cmdlet gets information about all the WSL distributions available from online sources.
```
Comment (SvenGroot) [inline docs/Get-WslDistributionOnline.md]: Please use a maximum line width of 100 characters in Markdown files.

File: docs/Get-WslDistributionOnline.md [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+SUSE-Linux-Enterprise-Server-15-SP4 SUSE Linux Enterprise Server 15 SP4
+SUSE-Linux-Enterprise-15-SP5        SUSE Linux Enterprise 15 SP5
+openSUSE-Tumbleweed                 openSUSE Tumbleweed
+```
+
+This example lists all WSL distributions available from online sources.
+
+### CommonParameters
+
+This cmdlet supports the common parameters: -Debug, -ErrorAction, -ErrorVariable, -InformationAction, -InformationVariable, -OutVariable, -OutBuffer, -PipelineVariable, -Verbose, -WarningAction, and -WarningVariable. For more information, see [about_CommonParameters](http://go.microsoft.com/fwlink/?LinkID=113216).
```
Comment (SvenGroot) [inline docs/Get-WslDistributionOnline.md:59]: This one's fine because it's generated by platyps.

File: docs/Get-WslDistributionOnline.md [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+None
+
+## OUTPUTS
+
+### WslDistributionOnline
+
+This cmdlet returns objects that represent the distributions available from online sources. This object has the
+following properties:
+
+- `Name`: The distribution name.
```
Comment (SvenGroot) [inline docs/Get-WslDistributionOnline.md]:
```suggestion
- `Name`: The distribution name that can be used to install it using `wsl.exe --install`.
```

File: docs/Get-WslDistributionOnline.md [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+
+[Import-WslDistribution](Import-WslDistribution.md)
+
+[Invoke-WslCommand](Invoke-WslCommand.md)
+
+[Remove-WslDistribution](Remove-WslDistribution.md)
+
+[Set-WslDistribution](Remove-WslDistribution.md)
+
+[Stop-WslDistribution](Remove-WslDistribution.md)
```
Comment (SvenGroot) [inline docs/Get-WslDistributionOnline.md]: These are not really related here; they're listed for Get-WslDistribution because they can operate on the output of that command, but that isn't the case here.

File: docs/Get-WslDistribution.md [hunk 16]
```diff
@@ -1,7 +1,7 @@
 ---
 external help file: Wsl-help.xml
 Module Name: Wsl
-online version: https://github.com/SvenGroot/WslManagementPS/blob/main/docs/Stop-WslDistribution.md
+online version: https://github.com/SvenGroot/WslManagementPS/blob/main/docs/Get-WslDistribution.md
```
Comment (SvenGroot) [inline docs/Get-WslDistribution.md:4]: Thanks for catching that!

**Review by SvenGroot — CHANGES_REQUESTED** [review]
Hi @JediNite, thank you so much for your interest in contributing to this project! I do have some comments on your change, but if you can address those I think this can be merged.

File: docs/Get-WslDistributionOnline.md (hunk 13 above)
Comment (JediNite) [reply docs/Get-WslDistributionOnline.md:59]: Noted.

File: docs/Get-WslDistributionOnline.md (hunk 12 above)
Comment (JediNite) [reply docs/Get-WslDistributionOnline.md]: Updated.

File: ChangeLog.md (hunk 11 above)
Comment (JediNite) [reply ChangeLog.md]: Updated.

File: docs/Get-WslDistributionOnline.md (hunk 15 above)
Comment (JediNite) [reply docs/Get-WslDistributionOnline.md]: Removed and updated with "None".

File: docs/Get-WslDistribution.md (hunk 16 above)
Comment (JediNite) [reply docs/Get-WslDistribution.md:4]: Noted.

File: docs/Get-WslDistributionOnline.md (hunk 14 above)
Comment (JediNite) [reply docs/Get-WslDistributionOnline.md]: Updated.

File: README.md (hunk 10 above)
Comment (JediNite) [reply README.md]: Updated.

File: Wsl.psd1 (hunk 8 above)
Comment (JediNite) [reply Wsl.psd1]: Updated.

File: Wsl.psm1 (hunk 7 above)
Comment (JediNite) [reply Wsl.psm1]: Updated.

File: Wsl.psm1 (hunk 6 above)
Comment (JediNite) [reply Wsl.psm1]: I will do some further testing with this.  When I tried doing this with Trim() previously, it did not remove the excess spaces at the start of the string effectively, and is why I resorted to the convoluted way in the current code, which removes all spaces and then reinserts them back in.

File: Wsl.psm1 (hunk 5 above)
Comment (JediNite) [reply Wsl.psm1]: Updated and added the comments just below the "elseif" start code statement.  Will do some testing with "-ceq" as well.  I think I tried using "-eq" and that did not match.

File: Wsl.psm1 (hunk 4 above)
Comment (JediNite) [reply Wsl.psm1]: Updated.

File: Wsl.psm1 (hunk 3 above)
Comment (JediNite) [reply Wsl.psm1]: Removed and will test, but suspect you are correct.

File: Wsl.psm1 (hunk 2 above)
Comment (JediNite) [reply Wsl.psm1]: Updated.  Still new to creating functions in this way.

File: Wsl.psm1 (hunk 1 above)
Comment (JediNite) [reply Wsl.psm1]: Updated.

File: Wsl.psm1 (hunk 5 above)
Comment (JediNite) [reply Wsl.psm1]: This seems to work based on testing.

File: Wsl.psm1 (hunk 6 above)
Comment (JediNite) [reply Wsl.psm1]: Trim() is also working correctly now.  This could be as a result of implementing the "-split" a few lines earlier as once I put this in did not go back and do more testing with the convoluted call.

File: Wsl.Tests.ps1 (hunk 9 above)
Comment (JediNite) [reply Wsl.Tests.ps1]: I've implemented that update, but not done a lot of Pester testing before, so will work out the required checks here.

Comment (JediNite) [reply Wsl.Tests.ps1]: Added a pester Test for this now.

**Comment by JediNite** [comment]
@SvenGroot,

All requested updated should now be present in commit [df315f9](https://github.com/SvenGroot/WslManagementPS/pull/8/commits/df315f9b562ba4d517951733e936e41d1e6609f6).

Cheers,
JediNite

**Comment by JediNite** [comment]
Also updated the references to "$IsWindows" in the module to "$IsWindowsOS" as i received a warning about using the value in PowerShell 6 or higher.

"Starting from PowerShell 6.0, the Variable 'IsWindows' cannot be assigned any more since it is a readonly automatic variable that is built into PowerShell, please use a different name.PSScriptAnalyzer(PSAvoidAssignmentToAutomaticVariable)"

File: Wsl.psm1 [hunk 17]
```diff
@@ -59,12 +59,18 @@ class WslVersionInfo {
...
+class WslDistributionOnline {
+    [string]$Name
+    [string]$FriendlyName
+}
+
+# Ensure IsWindowsOS is set for Windows PowerShell to make future checks easier.
 if ($PSVersionTable.PSVersion.Major -lt 6) {
-    $IsWindows = $true
+    $IsWindowsOS = $true
 }
```
Comment (SvenGroot) [inline Wsl.psm1]: Why did you change this? It breaks the module on PowerShell 6+.

File: ChangeLog.md [hunk 18]
```diff
@@ -1,5 +1,10 @@
 # What's new in the WSL Management PowerShell module
 
+## Version 2.1.0 (2023-09-26)
```
Comment (SvenGroot) [inline ChangeLog.md]:
```suggestion
## Version 2.1 (2023-09-26)
```

File: docs/Get-WslDistributionOnline.md [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+OracleLinux_9_1                     Oracle Linux 9.1
+openSUSE-Leap-15.5                  openSUSE Leap 15.5
+SUSE-Linux-Enterprise-Server-15-SP4 SUSE Linux Enterprise Server 15 SP4
+SUSE-Linux-Enterprise-15-SP5        SUSE Linux Enterprise 15 SP5
+openSUSE-Tumbleweed                 openSUSE Tumbleweed
+```
+
+This example lists all WSL distributions available from online sources.
+
+### CommonParameters
```
Comment (SvenGroot) [inline docs/Get-WslDistributionOnline.md:57]: New-ExternalHelp fails to process this file for a few reasons. Please run Create-Release.ps1 to ensure there are no errors.

The first problem is that you are missing the `## PARAMETERS` heading before this heading.

File: docs/Get-WslDistributionOnline.md [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+
+This cmdlet supports the common parameters: -Debug, -ErrorAction, -ErrorVariable, -InformationAction, -InformationVariable, -OutVariable, -OutBuffer, -PipelineVariable, -Verbose, -WarningAction, and -WarningVariable. For more information, see [about_CommonParameters](http://go.microsoft.com/fwlink/?LinkID=113216).
+
+## INPUTS
+
+None
```
Comment (SvenGroot) [inline docs/Get-WslDistributionOnline.md]: You can't have `None` here; just leave the section empty (see e.g. Get-WslVersion.md).

File: docs/Get-WslDistributionOnline.md [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+object has the following properties:
+
+- `Name`: The distribution name that can be used to install it using `wsl.exe --install`
+- `FriendlyName`: The Friendly Name for the distribution.
+
+## NOTES
+
+## RELATED LINKS
+
+None
```
Comment (SvenGroot) [inline docs/Get-WslDistributionOnline.md]: You can't have `None` here either. While you can leave it empty, I think having a related link to only Get-WslDistribution is better.

File: ChangeLog.md [hunk 22]
```diff
@@ -1,5 +1,10 @@
...
+
+- Added support for retrieving WSL distribution listing from online sources via
+  [`Get-WslDistributionOnline`][] cmdlet.
```
Comment (SvenGroot) [inline ChangeLog.md:6]: There is no matching link for this label, please add one.

File: Wsl.psm1 (hunk 17 above)
Comment (SvenGroot) [reply Wsl.psm1]: Just saw your other comment about this. Yes, you get a warning here, but it's a false positive because the code only assigns the variable on PowerShell older than 6. The intent here is to use the built-in variable in PowerShell 6+, and just set it to true on Windows PowerShell 5.1 because we know that's always Windows. Please revert this change.

**Review by SvenGroot — CHANGES_REQUESTED** [review]
@JediNite Thanks for making those changes. I did some verification and ran into a few issues, which I've left additional comments for.

File: Wsl.psm1 (hunk 17 above)
Comment (JediNite) [reply Wsl.psm1]: Reverted.

File: docs/Get-WslDistributionOnline.md (hunk 21 above)
Comment (JediNite) [reply docs/Get-WslDistributionOnline.md]: Updated

File: ChangeLog.md (hunk 18 above)
Comment (JediNite) [reply ChangeLog.md]: Updated.

File: ChangeLog.md (hunk 22 above)
Comment (JediNite) [reply ChangeLog.md:6]: Updated.  Sorry never done links in Markdown before so had to work it out. :)

File: docs/Get-WslDistributionOnline.md (hunk 20 above)
Comment (JediNite) [reply docs/Get-WslDistributionOnline.md]: Updated

File: docs/Get-WslDistributionOnline.md (hunk 19 above)
Comment (JediNite) [reply docs/Get-WslDistributionOnline.md:57]: Added ##PARAMETERS heading and ran Create-Release.ps1 which created output in release\Wsl directory that at first glance looked okay.  Please confirm.

**Review by SvenGroot — APPROVED** [review]
Okay, this looks good to me now. Thanks again for your contribution! I'll try to get an updated version out soon.

**Comment by SvenGroot** [comment]
@JediNite Version 2.1 is [now available](https://www.powershellgallery.com/packages/Wsl/2.1.0). Thanks again!
